### PR TITLE
Implement CLI methods for `connect` and `disconnect`

### PR DESCRIPTION
**Pull request by claddyk**
### This PR addresses issue #6 

- [x] Add the `connect` Command.
- [x] Add the `disconnect` Command.
- [x] Unit Tests for `connect`
- [x] Unit Tests for `disconnect`
- [ ] Work on [this comment](https://github.com/ACINQ/eclair-cli/pull/9#discussion_r1273186358)

### CODE REVIEW

**Comment by claddyk** [comment]
The only thing left in this PR is to solve this [issue](https://github.com/ACINQ/eclair-cli/pull/9#discussion_r1282830282).

**Comment by claddyk** [comment]
Since we have the [Base command](https://github.com/ACINQ/eclair-cli/pull/12/files#diff-c61834c848eb6e1b45ae0e1f0d14467630c6a180a7763862ab25d73fd76dcbb6) now, shall I implement the same changes in the [GetInfo](https://github.com/ACINQ/eclair-cli/blob/master/src/nativeMain/kotlin/commands/GetInfo.kt) command too?

File: src/nativeMain/kotlin/api/EclairClient.kt [hunk 1]
```diff
@@ -22,6 +24,14 @@ class EclairClientBuilder : IEclairClientBuilder {
 
 interface IEclairClient {
     suspend fun getInfo(): Either<ApiError, String>
+    sealed class ConnectionTarget {
```
Comment (t-bast) [inline src/nativeMain/kotlin/api/EclairClient.kt:27]: Can you move this `sealed class` outside of the `IEclairClient` interface?

File: src/nativeMain/kotlin/api/EclairClient.kt [hunk 2]
```diff
@@ -51,11 +69,55 @@ class EclairClient(private val apiHost: String, private val apiPassword: String)
...
+                        append("address", target.address)
+                        target.port?.let { append("port", it.toString()) }
+                    }
+                )
+            }
+            when (response.status) {
+                HttpStatusCode.OK -> Either.Right(Json.decodeFromString(response.bodyAsText()))
+                else -> Either.Left(convertHttpError(response.status))
+            }
+        } catch (e: Exception) {
```
Comment (t-bast) [inline src/nativeMain/kotlin/api/EclairClient.kt:102]: nit: you should catch `Throwable`, which is a superset of `Exception`

File: src/nativeMain/kotlin/commands/Connect.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import kotlinx.coroutines.runBlocking
+
+class ConnectCommand(
+    private val resultWriter: IResultWriter,
+    private val eclairClientBuilder: IEclairClientBuilder
+) : BaseCommand(
+    "connect",
+    "Connect to another lightning node. This will perform a connection but no channels will be opened. "
+) {
+    var uri by option(
```
Comment (t-bast) [inline src/nativeMain/kotlin/commands/Connect.kt:16]: All of those fields should be `private val`, not `var`

File: src/nativeMain/kotlin/commands/Connect.kt [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        description = "Connect to another lightning node. This does not require a target address. Instead, eclair will use one of the addresses published by the remote peer in its node_announcement messages."
+    )
+
+    var address by option(
+        ArgType.String,
+        description = "The IPv4 host address of the node."
+    )
+    var port by option(
+        ArgType.Int,
+        description = "The port of the node(default: 9735)"
```
Comment (t-bast) [inline src/nativeMain/kotlin/commands/Connect.kt:32]: nit:
```suggestion
        description = "The port of the node (default: 9735)"
```

File: src/nativeMain/kotlin/commands/Connect.kt [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            nodeId != null && address == null -> eclairClient.connect(IEclairClient.ConnectionTarget.NodeId(nodeId!!))
+            nodeId != null && address != null -> eclairClient.connect(
+                IEclairClient.ConnectionTarget.Manual(
+                    nodeId!!,
+                    address!!,
+                    port
+                )
+            )
+
+            else -> throw IllegalArgumentException("Either URI or nodeId must be provided.")
```
Comment (t-bast) [inline src/nativeMain/kotlin/commands/Connect.kt:48]: As a general rule, you should avoid throwing exceptions. You really don't need it, this is an error case where you simply need to write an error to the `resultWriter`.

File: src/nativeMain/kotlin/commands/Disconnect.kt [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package commands
+
+import IResultWriter
+import api.IEclairClientBuilder
+import kotlinx.cli.ArgType
+import kotlinx.cli.ExperimentalCli
+import kotlinx.cli.required
+import kotlinx.coroutines.runBlocking
+
+@OptIn(ExperimentalCli::class)
```
Comment (t-bast) [inline src/nativeMain/kotlin/commands/Disconnect.kt:10]: This is unnecessary, remove?

**Comment by t-bast** [comment]
> The only thing left in this PR is to solve this https://github.com/ACINQ/eclair-cli/pull/9#discussion_r1282830282.

Yes, once you've fixed the other comments in this PR, you should define real data classes for the successful results of the `connect` and `disconnect` APIs.

**Comment by claddyk** [comment]
> > The only thing left in this PR is to solve this [#9 (comment)](https://github.com/ACINQ/eclair-cli/pull/9#discussion_r1282830282).
> 
> Yes, once you've fixed the other comments in this PR, you should define real data classes for the successful results of the `connect` and `disconnect` APIs.

I've shared my approach on this [here](https://github.com/ACINQ/eclair-cli/pull/9#discussion_r1282830282).
Can you please provide little bit of code for more clarity on it.

**Comment by t-bast** [comment]
> Can you please provide little bit of code for more clarity on it.

Sure, you simply need to define a type for the result of the connection (in `EclairApiTypes.kt`):

```kotlin
@Serializable
data class ConnectionResult(val success: Boolean) : EclairApiType()
```

Then in the `execute()` of the `Connect` command, you convert the result to your `ConnectionResult` and serialize it:

```kotlin
        val result = when {
            uri != null -> ...
            ...
        }.map { response ->
            val result = when (response) {
                "connected" -> ConnectionResult(true)
                else -> ConnectionResult(false)
            }
            Serialization.encode(result)
        }
```

**Comment by claddyk** [comment]
> > Can you please provide little bit of code for more clarity on it.
> 
> Sure, you simply need to define a type for the result of the connection (in `EclairApiTypes.kt`):
> 
> ```kotlin
> @Serializable
> data class ConnectionResult(val success: Boolean) : EclairApiType()
> ```
> 
> Then in the `execute()` of the `Connect` command, you convert the result to your `ConnectionResult` and serialize it:
> 
> ```kotlin
> val result = when {
>     uri != null -> ...
>     ...
> }.map { response ->
>     val result = when (response) {
>         "connected" -> ConnectionResult(true)
>         else -> ConnectionResult(false)
>     }
>     Serialization.encode(result)
> }
> ```

The logs after implementing this are:
- On success
```
claddy@claddy:~/Documents/eclair-cli$ eclair-cli.kexe connect -p password --uri 039743b8b5030cf15717f855e7a4e4d15b40c0c08ed2f219c739721570baae81d5@127.0.0.1:9737
{
  "type": "types.ConnectionResult",
  "success": true
}
```
- On failure
```
claddy@claddy:~/Documents/eclair-cli$ eclair-cli.kexe connect -p password --uri 02f666711319435b
7905dd77d10c269d8d50c02668b975f526577167d370b50a3e@127.0.0.1:9736
{
  "type": "types.ConnectionResult",
  "success": false
}
```
The old eclair-cli returns `already connected` for this case.
```
claddy@claddy:~/Documents/eclair-cli$ eclair-cli connect --uri=02f666711319435b7905dd77d10c269d8d50c02668b975f526577167d370b50a3e@127.0.0.1:9736
Enter host password for user 'eclair-cli':
already connected
```
Is it fine, if we put it in the category of false?
Technically it makes sense, but might confuse the user.

**Comment by t-bast** [comment]
> Is it fine, if we put it in the category of false?
> Technically it makes sense, but might confuse the user.

If we're already connected, this should return `true`, that's what makes the most sense: the `connect` API was a success since we're connected.

**Comment by claddyk** [comment]
Please ignore commit ea92360880f1b5958f89981b265783198adee2e2 as of now.
I'll ammend it properly.

**Review by t-bast — APPROVED** [review]
LGTM :+1:

**Pull request merged**